Feasibility pump and updates in GOA

ZedongPeng opened the pull request:
## Fixes # .

## Summary/Motivation:


## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

bernalde commented on pyomo/contrib/mindtpy/initialization.py:
```diff
@@ -136,10 +139,12 @@ def init_rNLP(solve_data, config):
-            solve_data.LB = results['Problem'][0]['Lower bound']
-        elif not math.isnan(results['Problem'][0]['Upper bound']):
-            solve_data.UB = results['Problem'][0]['Upper bound']
+        # TODO: should we use the bound of the rNLP here?
```
If the solver provides a lower bound, we could use it (in the case for example we do not trust the optimal solution of the rNLP). I think that in general, we decide to trust the bound though.

bernalde commented on pyomo/contrib/mindtpy/mip_solve.py:
```diff
@@ -148,14 +148,14 @@ def solve_master(solve_data, config, feas_pump=False):
                 solve_data.mip, tee=config.mip_solver_tee, **mip_args)
 
-        # if config.single_tree is False and config.add_nogood_cuts is False:
+        # if config.single_tree is False and config.add_no_good_cuts is False:
```
Do we still need this comment? What was its original aim?

bernalde commented on pyomo/contrib/mindtpy/nlp_solve.py:
```diff
@@ -48,7 +48,10 @@ def solve_subproblem(solve_data, config):
     TransformationFactory('core.fix_integer_vars').apply_to(fixed_nlp)
 
     MindtPy.MindtPy_linear_cuts.deactivate()
+    # TODO: Do we really need to calculate tmp_duals by default?
```
This is a good point. If we do not add the equality relaxation cuts this is time-consuming and dumb. Consider removing it.

bernalde commented on pyomo/contrib/mindtpy/nlp_solve.py:
```diff
@@ -71,7 +74,8 @@ def solve_subproblem(solve_data, config):
                 0, c_geq*(rhs - value(c.body)))
-        except (ValueError, OverflowError) as error:
+        except (ValueError, OverflowError, AttributeError) as error:
+            # AttributeError is included to condition when value(c.body) belongs to complex type
```
When does this happen!?

ZedongPeng replied:
This happens when `value(c.body)` is a complex number. In some examples, due to the numerical computation, `1+0.0001j` may show up.

bernalde commented on pyomo/contrib/mindtpy/nlp_solve.py:
```diff
@@ -185,12 +195,12 @@ def handle_subproblem_optimal(fixed_nlp, solve_data, config, feas_pump=False):
         add_affine_cuts(solve_data, config)
     elif config.strategy == 'PSC':
```
If we are commenting GDP we should also comment PSC

bernalde commented on pyomo/contrib/mindtpy/nlp_solve.py:
```diff
@@ -354,17 +366,24 @@ def solve_feasibility_subproblem(solve_data, config):
     elif subprob_terminate_cond in {tc.infeasible, tc.noSolution}:
-        raise ValueError('Feasibility NLP infeasible. '
-                         'This should never happen.')
+        config.logger.error('Feasibility NLP infeasible. '
```
Feasibility subproblem

bernalde commented on pyomo/contrib/mindtpy/single_tree.py:
```diff
@@ -241,14 +232,13 @@ def add_lazy_affine_cuts(self, solve_data, config, opt):
             config.logger.info("Added %s affine cuts" % counter)
 
-    def add_lazy_nogood_cuts(self, var_values, solve_data, config, opt, feasible=False):
+    def add_lazy_no_good_cuts(self, var_values, solve_data, config, opt, feasible=False):
```
I thought there was a TODO above mentioning this. Are we using no-good in single tree? Did you try the taboo list mentioned in the storm of feasibility pump paper?
A way of implementing this is their section 5.2 using https://www.ibm.com/support/knowledgecenter/SSSA5P_12.10.0/ilog.odms.cplex.help/refpythoncplex/html/cplex.callbacks.Context-class.html#reject_candidate

bernalde commented on pyomo/contrib/mindtpy/tests/test_mindtpy_feas_pump.py:
```diff
@@ -23,7 +23,7 @@
-required_solvers = ('ipopt', 'cplex')
+required_solvers = ('ipopt', 'glpk', 'cplex')
 if all(SolverFactory(s).available() for s in required_solvers):
     subsolvers_available = True
```
This will skip some of the tests of CPLEX is not available although mots of them run with GLPK.

bernalde commented on pyomo/contrib/mindtpy/tests/test_mindtpy_lp_nlp.py:
```diff
@@ -158,7 +158,7 @@ def test_OA_OnlineDocExample(self):
     #                   mip_solver=required_solvers[1],
     #                   nlp_solver=required_solvers[0],
-    #                   add_nogood_cuts=True,
+    #                   add_no_good_cuts=True,
```
Why is this test still commented?

bernalde reviewed (commented):
Great PR. I just left some minor comments and a question open on how to implement the tabu list from the storms of feasibility pumps (according to the paper it is WAY better than no good cuts, so we might give it a look).
![image](https://user-images.githubusercontent.com/8647329/97509381-82f76580-1958-11eb-92fd-c49b88f9c83c.png)

ZedongPeng commented:
> Great PR. I just left some minor comments and a question open on how to implement the tabu list from the storms of feasibility pumps (according to the paper it is WAY better than no good cuts, so we might give it a look).
> ![image](https://user-images.githubusercontent.com/8647329/97509381-82f76580-1958-11eb-92fd-c49b88f9c83c.png)

I am a little confused about how they handle an integer solution found by CPLEX at the root node appears in the tabu list. Will the root node make a difference?

blnicho committed:
Merge branch 'main' into replace-simple

jsiirola committed:
Removing pyomo.core.preprocess.simple_preprocessor

I am pretty sure this had a syntax error (undefined rank variable) and
was never used (certainly not covered by any tests).

jsiirola committed:
Removing pyomo_api and the pyutilib workflow dependency

jsiirola committed:
Allow assigning None to a united variable

jsiirola committed:
NFC: fixing doc typos

blnicho committed:
Merge branch 'main' into update-book-tests

blnicho committed:
Adding baseline file for additional GDP test

blnicho committed:
Merge pull request #1960 from jsiirola/disjunct-bool-indicator-var

Convert `Disjunct.indicator_var` from binary to Boolean variable

blnicho committed:
Skipping GDP book test when sympy is not available

jsiirola committed:
Merge pull request #1977 from jsiirola/assign-none-to-united-var

Allow assigning None to a unitted variable

Commit:
Merge remote-tracking branch 'upstream/main' into network_patch

jsiirola committed:
Make test more deterministic on pypy

jsiirola committed:
NFC: fix doc typos

blnicho committed:
Merge pull request #1914 from blnicho/update-book-tests

Update Pyomo book examples & tests for the 3rd edition

michaelbynum committed:
Merge pull request #1961 from bknueven/network_patch

pyomo.network / pyomo.gdp interoperability

michaelbynum committed:
Merge branch 'main' into ineqs2

jsiirola committed:
Merge branch 'main' into rework-pyomo-script

mrmundt committed:
Merging main and renaming Simples in base.set

mrmundt committed:
Merge branch 'replace-simple' of github.com:mrmundt/pyomo into replace-simple

mrmundt committed:
Replacing remaining instances of Simple

jsiirola committed:
Removing unneeded code; restoring issuperset(_NonHashable) test

blnicho committed:
Merge branch 'replace-simple' of https://github.com/mrmundt/pyomo into replace-simple

blnicho committed:
Replacing use of SimpleSet in contset.py

jsiirola committed:
Fixing incorrect conflict resolution from cc33f67

jsiirola committed:
Acknowledge the origin of the plugin system

mrmundt committed:
More replacement of Single with Scalar

mrmundt committed:
Merge branch 'replace-simple' of github.com:mrmundt/pyomo into replace-simple

jsiirola committed:
Promote assertStructureAlmostEqual to a module function

This supports use outside of unittest.  Also, add an item_callback
option so the caller can provide a callback for converting data items
before comparison.

mrmundt committed:
Removing residual references to pyutilib / adding soft dependency

jsiirola committed:
NFC: fix doc typo

mrmundt committed:
Changing init to check for pyutilib

blnicho committed:
Merge pull request #1847 from ZedongPeng/new-lp-nlp

Updates in MindtPy

mrmundt committed:
Changing skipIf and adding pyutilib back into environ timing check

mrmundt committed:
Removing pyomo_python console script

mrmundt committed:
Updating language in contribution guide

mrmundt committed:
Merge pull request #1976 from jsiirola/rework-pyomo-script

Remove PyUtilib dependencies from pyomo script

mrmundt committed:
Fixing readthedocs rendering

mrmundt committed:
Fixing typo

mrmundt committed:
Removing pyomo_uninstall script

mrmundt committed:
Merge branch 'main' into divorce-pyutilib

mrmundt committed:
Pyutilib.th was being used on main

mrmundt committed:
Merging and fixing conflicts

mrmundt committed:
Removing six imports

michaelbynum committed:
utilities to compare expressions

michaelbynum committed:
utilities to compare expressions

michaelbynum committed:
utilities to compare expressions

mrmundt committed:
Updating test_environ TPLs and removing Pyro4 instances

mrmundt committed:
Reinstating decimal as slow TPL

mrmundt committed:
Adding a few more TPLs

mrmundt committed:
Rearranging TPLs

blnicho committed:
Merge pull request #1954 from michaelbynum/ineqs2

Only cast constant expressions to bool

blnicho committed:
Merge pull request #1971 from mrmundt/replace-simple

Replace `Simple` with `Scalar`

jsiirola committed:
Merge branch 'main' into expr_comp

jsiirola committed:
PEP8 formatting

jsiirola committed:
Simplifying walker logic to remove redundant code

jsiirola committed:
Merge branch 'main' into assert-structured-fcn

blnicho committed:
Merge pull request #1981 from michaelbynum/expr_comp

Utility for comparing expressions

blnicho committed:
Merge pull request #1979 from jsiirola/assert-structured-fcn

Promote `assertStructuredAlmostEqual` to a module function

jsiirola committed:
More robust quoting of component names/indices.

This provides a more robust generation of component names (a slight
generalization of what was developed for ComponentUID) that reliably
reports length-1 tuples as tuples and quotes names with special tokens.

jsiirola committed:
Improve collision detection in short name labeler

jsiirola committed:
Preserve the local name when creating concrete instances

jsiirola committed:
Updating tests to reflect more robust name quoting

mrmundt committed:
Minor changes

mrmundt committed:
Merge branch 'main' into divorce-pyutilib

blnicho committed:
Updating CHANGELOG for the 6.0 release

blnicho committed:
Merge branch 'main' into finalize-release

mrmundt committed:
Adding fail-fast nosetest option to pyomo unittest

blnicho committed:
More updates to the CHANGELOG for the 6.0 release

mrmundt committed:
Fixing nosetests executable discovery

jsiirola committed:
Updating tests to reflect quoting in fully-qualified names

jsiirola committed:
Remove the implementation of reconstruct from the Component API

jsiirola committed:
Remove use of reconstruct() from pyomo.dae

blnicho committed:
Updating filter for fragile performance-ch book test

jsiirola committed:
Fix exception instantiating IndexedConstraint with expr

jsiirola committed:
Defining clear() for scalar Objective, Expression, and Constraint

jsiirola committed:
NFC: updating comment

jsiirola committed:
Removing debugging

jsiirola committed:
Define clear() for ScalarInterval

jsiirola committed:
Support 'reconstruction' of scalar Objectives

jsiirola committed:
Improve diagnostic output

mrmundt committed:
Merge pull request #1980 from mrmundt/divorce-pyutilib

Remove hard `PyUtilib` dependency

blnicho committed:
Merge pull request #1982 from jsiirola/quote-component-names

Improve quoting in component names and indices

jsiirola committed:
Move to using 'yield from' in kernel generators

jsiirola committed:
Adding a pypy build to the branch tests

jsiirola committed:
Additional moves to using 'yield from' in core, repn

blnicho committed:
Merge branch 'main' into finalize-release

blnicho committed:
More updates to the CHANGELOG for the 6.0 release

blnicho committed:
Merge branch 'main' into finalize-release

blnicho committed:
More updates to the CHANGELOG for the 6.0 release

blnicho committed:
Merge pull request #1983 from jsiirola/remove-reconstruct

Remove Component.reconstruct()

jsiirola committed:
Merge branch 'main' into kernel-generators

blnicho committed:
More updates to the CHANGELOG for the 6.0 release

jsiirola committed:
Update name generator/parser to better handle special_characters

jsiirola committed:
Removing debugging

blnicho committed:
Merge pull request #1985 from jsiirola/spaces-in-component-names

Update name generator/parser to better handle special_characters

michaelbynum committed:
ensuring appsi tests get skipped if extensions are not built

michaelbynum committed:
slim builds should not install extensions

michaelbynum committed:
slim builds should not install extensions

mrmundt committed:
Merge pull request #1988 from michaelbynum/appsi_tests

Skip ppsi tests and Modify slim-build

jsiirola committed:
Merge branch 'main' into finalize-release

jsiirola committed:
Updating the changelog

jsiirola committed:
Setting deprecation warning version number to 6.0

jsiirola committed:
Updating version to 6.0

jsiirola committed:
Update setup to include extensions source files

jsiirola committed:
Merge pull request #1989 from Pyomo/finalize-release

Finalize 6.0 release

jsiirola committed:
Preparing master for development (6.0.1.dev0)

jsiirola committed:
Merge pull request #1990 from Pyomo/finalize-release

Preparing master for development (6.0.1.dev0)

jsiirola committed:
Clean up logic for skipping extensions on slim builds

jsiirola committed:
Enable AmplInterface.available() even in the absence of numpy

jsiirola committed:
Skip some incidence_analysis tests when not AmplInterface.available()

jsiirola committed:
Switch piecewise solve tests to use gurobi

jsiirola committed:
Update baseline string names (quoting strings)

jsiirola committed:
Fix regression importing numpy

emma58 committed:
Using fbbt to calculate M for linear expressions too

emma58 committed:
Remove extra space in error message

michaelbynum committed:
fbbt support for linear expression

ZedongPeng committed:
add unit_test.py

ZedongPeng committed:
increase code coverage

michaelbynum committed:
Merge pull request #1994 from michaelbynum/fbbt_linear_expr

fbbt support for linear expression

jsiirola committed:
Catch non-standard file.write() implementations

Catch non-standard file object implementations that do not return the
number of characters written as the result of `write()`.  Fixes #1997.

jsiirola committed:
Update expected test output

jsiirola committed:
Merge pull request #1991 from jsiirola/testing-fixes

Testing fixes and updates

jsiirola committed:
Add explanatory comment

jsiirola committed:
Merge pull request #1999 from jsiirola/tee-ipython-fix

Catch non-standard `file.write()` implementations

emma58 committed:
Merge branch 'main' into switch-bigm-to-fbbt

michaelbynum committed:
Merge pull request #2000 from emma58/switch-bigm-to-fbbt

Switch bigm's M value calculation to always use fbbt

jsiirola committed:
Implement redirect_fd for robust redirection of file descriptors

jsiirola committed:
Move cyipopt to using common.tee.redirect_fd

jsiirola committed:
Fixing typo (undefined variable)

jsiirola committed:
Fixing doc typo

jsiirola committed:
Add redirect_fd documentation; remove enable flag

jsiirola committed:
Fix redirection of cyipopt stdout to sys.stdout

jsiirola committed:
Fixing typo (copy/paste error)

jsiirola committed:
Removing deprecated import of 'ipopt' module

jsiirola committed:
Update cyipopt interface to use the 1.0.0 API

jsiirola committed:
NFC: fix doc typo

jsiirola committed:
Ensure test output is flushed (fixes failures on pypy)

blnicho committed:
A few minor changes in GDPopt and examples to avoid implicit conversion warnings

blnicho committed:
Updating the CHANGELOG in preparation for the 6.0.1 release

jsiirola committed:
Silence deprecation warnings in examples/gdp/stickies.py

blnicho committed:
Merge pull request #2001 from jsiirola/capture-fd

Improve management of stdout in cyipopt solver

blnicho committed:
Merge branch 'main' into finalize-release-6.0.1

blnicho committed:
Merge pull request #2002 from blnicho/fix-gdpopt-examples

Fixes to GDPopt and examples to avoid implicit conversion warnings

blnicho committed:
More updates to the CHANGELOG for the 6.0.1 release

blnicho committed:
Merge branch 'main' into finalize-release-6.0.1

jsiirola committed:
Updating version to 6.0.1

blnicho committed:
Merge pull request #2003 from blnicho/finalize-release-6.0.1

Finalize release 6.0.1

jsiirola committed:
Resetting main for development (6.0.2.dev0)

blnicho committed:
Merge pull request #2004 from blnicho/finalize-release-6.0.1

Resetting main for development (6.0.2.dev0)

mrmundt committed:
Changing implementation of keys, values, items in IndexedComponent

mrmundt committed:
Attempting use of iter option

mrmundt committed:
Converting test generator to list

mrmundt committed:
Changing cast

mrmundt committed:
Converting linear_coefs to a list

mrmundt committed:
Merge pull request #2007 from mrmundt/iterators

Yield Iterator in `IndexedComponent`

jsiirola committed:
Allow objects other than strings as attempt_import(error_message)

jsiirola committed:
LoggingIntercept: log to an implicit StringIO if no buffer provided

jsiirola committed:
Remove explicit dependency on parameterized in APPSI

jsiirola committed:
cyipopt.available() should always return a bool

jsiirola committed:
Remove hard dependency on nose

jsiirola committed:
Ensure direct/persistent solver .available() always returns bool

jsiirola committed:
error_message=None should be handled the same as an empty string

jsiirola committed:
Capture stdout when creating gurobi model to test license

jsiirola committed:
Add a backport of the Python 3.7 nullcontext

jsiirola committed:
Suppress executable not found error in SystemCallSolver.available()

jsiirola committed:
Update Gurobi direct/persistent to leverage attempt_import

Also, suppress stdout when testing available()

jsiirola committed:
Update xpress_direct to leverage attempt_import

Also suppress stdout within available()

jsiirola committed:
Add a standalone Pyomo test driver

This script exercises a writer, creates (almost) all solvers and checks
their availability (ensuring that `available(False)` is silent: no
stdout, no logged messages), creates all (nondeprecated) transformations
and checks that they can be created silently.

jsiirola committed:
Adding '[DEPRECATED]' to deprecated transformation documentation

jsiirola committed:
Add a test that Pyomo installs/runs in a "bare" Python environment

jsiirola committed:
Fixing typo (trailing ")

jsiirola committed:
Fix parameterized import

jsiirola committed:
Simplify import

jsiirola committed:
Remove instance-level definition of _version_major

jsiirola committed:
Fix instance definition of _version attribute

jsiirola committed:
Fix masking _version in gurobi, xpress direct solvers

jsiirola committed:
Fix typo (calling method on class)

jsiirola committed:
Remove pyutilib from branch tests; add bare to pr tests

jsiirola committed:
Support using the TicTocTimer as a context manager

jsiirola committed:
Relax timing tests on pypy

jsiirola committed:
Define 'tests' and 'docs' install dependencies

jsiirola committed:
Switch RTD to using a .readthedocs.yaml configuration

Dependencies are now set in setup.py and installed using
`pip install '.[docs]'`

jsiirola committed:
Relaxing the timing resolution

jsiirola committed:
Adding additional dependencies needed for RTD

jsiirola committed:
Switch timing tests to using the perf_counter (for accuray)

jsiirola committed:
Improve robustness of timing test

michaelbynum committed:
Merge pull request #2012 from jsiirola/timing-updates

Support using TicTocTimer as a context manager

jsiirola committed:
Adding mixxing copyright; fixing comment typos

jsiirola committed:
changing step order in bare-env test

jsiirola committed:
Merge pull request #2011 from jsiirola/dependency-andsolverio-cleanup

Clean up dependencies and `solver.available()` IO

blnicho committed:
Merge pull request #1984 from jsiirola/kernel-generators

[WIP] Use `yield from` in `pyomo.core`

jsiirola committed:
Adding [optional] installation target

jsiirola committed:
Update patch to suppress building benchmark and nl-example targets

These do not cleanly build on newer compilers and are not needed for libasl.

jsiirola committed:
Switch GHA to ubuntu-latest

jsiirola committed:
(debug) add pypy to the GHA branch build

jsiirola committed:
Restore spaces in patch (required for windows patch)

jsiirola committed:
Restore branch linux/3.9 build

jsiirola committed:
Further relax timing targets on pypy

jsiirola committed:
(debug) make parmest more verbose

jsiirola committed:
(debug) print out PATH before running tests

jsiirola committed:
(debug) print out .profile

jsiirola committed:
Move local runner paths before conda

jsiirola committed:
Update PATH in both .profile & .bash_profile

jsiirola committed:
Attempt to correct escaping

jsiirola committed:
ensuring a CR after conda config

jsiirola committed:
Removing debugging

jsiirola committed:
Synchronizing branch/pr workflows

jsiirola committed:
removing debugging

jsiirola committed:
Add 'setup.py dependencies' command

jsiirola committed:
Switch branch builder to use setup.py for dependencies

jsiirola committed:
Merge pull request #2015 from jsiirola/pynumero-build-updates

Update GHA to ubuntu-latest; resolve ampl/mp build error

jsiirola committed:
Fixing typo

jsiirola committed:
Fix catch of pip install

jsiirola committed:
Add fallback for older setuptools

jsiirola committed:
Invoke pip with PYTHON_EXE

jsiirola committed:
Correct fallback distutils import

jsiirola committed:
Do not use 'conda develop': it does not appear to handle console_scripts

jsiirola committed:
Allow empty 'extrs' argument

jsiirola committed:
Filter out packages not available through conda

jsiirola committed:
Always use 'pip install .[extras]' to install pyomo

jsiirola committed:
(debug) quoting dependencies

jsiirola committed:
(debug) pip command arguments

jsiirola committed:
remove use of pip -e

jsiirola committed:
Move pyomo dependency installs earlier

jsiirola committed:
call python directly

jsiirola committed:
Update dependency management

jsiirola committed:
Defer dependencies until after Python is set up

jsiirola committed:
typo: quoting

jsiirola committed:
remove package quoting

jsiirola committed:
Merge branch 'main' into setup-updates

jsiirola committed:
Explicitly list packages that are only available on PyPI

jsiirola committed:
resynchronixe pr/main builder

jsiirola committed:
typo: 'pypi'

jsiirola committed:
Remove extra backquote

jsiirola committed:
Capitalize DependenciesCommand class; clarify documentation

jsiirola committed:
Fix typos, remove unneeded commented out code

jsiirola committed:
Resolve minor doctest errors; add +IGNORE_RESULT option

jsiirola committed:
Make more optional dependencies available on pypy

jsiirola committed:
avoid direct execution of 'pip'

jsiirola committed:
Avoiding casadi under pypy

jsiirola committed:
Avoiding pandas under pypy

jsiirola committed:
Resyncing build scripts

jsiirola committed:
Bumping cache version

jsiirola committed:
Avoiding numdifftools, pandas under pypy

jsiirola committed:
Install testing dependencies when processing coverage

jsiirola committed:
Move constrained_layout tests to 'expensive'; add expensive coverage

mrmundt committed:
Merge pull request #2017 from jsiirola/setup-updates

Embed dependencies in setup.py

mrmundt committed:
Merge pull request #2018 from jsiirola/constrained-layout

Move constrained_layout tests to 'expensive' suite

jsiirola committed:
Merge branch 'main' into new-lp-nlp

bernalde committed:
Merge pull request #17 from ZedongPeng/new-lp-nlp

Implementation of LOA single-tree
